llvm/include/llvm/CAS/CASReference.h

```diff
@@ -0,0 +1,278 @@
+//===- llvm/CAS/CASReference.h ----------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CAS_CASREFERENCE_H
+#define LLVM_CAS_CASREFERENCE_H
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/DenseMapInfo.h"
+#include "llvm/ADT/PointerIntPair.h"
+#include "llvm/ADT/StringRef.h"
+
+namespace llvm {
+
+class raw_ostream;
+
+namespace cas {
+
+class CASDB;
+
+namespace testing_helpers {
+class HandleFactory;
+}
+
+/// Base class for references to things in \a CASDB.
+class ReferenceBase {
+public:
+  /// Get an internal reference.
+  uint64_t getInternalRef(const CASDB &ExpectedCAS) const {
+#if LLVM_ENABLE_ABI_BREAKING_CHECKS
+    assert(CAS == &ExpectedCAS && "Extracting reference for the wrong CAS");
+#endif
+    return InternalRef;
+  }
+
+  /// Print internal ref and/or CASID. Only suitable for debugging.
+  void print(raw_ostream &OS) const;
+
+  LLVM_DUMP_METHOD void dump() const;
+
+protected:
+  bool isEqual(const ReferenceBase &RHS) const {
+#if LLVM_ENABLE_ABI_BREAKING_CHECKS
+    assert(CAS == RHS.CAS && "Cannot compare across CAS instances");
+    assert(IsHandle == RHS.IsHandle && "Cannot compare handles to refs");
+#endif
+    return InternalRef == RHS.InternalRef;
+  }
+
+protected:
+  friend class CASDB;
+  friend class testing_helpers::HandleFactory;
+  ReferenceBase(const CASDB *CAS, uint64_t InternalRef, bool IsHandle)
+      : InternalRef(InternalRef) {
+#if LLVM_ENABLE_ABI_BREAKING_CHECKS
+    this->CAS = CAS;
+    this->IsHandle = IsHandle;
+#endif
+  }
+
+public:
+  enum HandleKind {
+    TreeKind = 0x1,
+    NodeKind = 0x2,
+    BlobKind = 0x4,
+    AnyDataKind = NodeKind | BlobKind,
+    AnyObjectKind = AnyDataKind | TreeKind,
+  };
+
+private:
+  friend class AnyObjectHandle;
+  template <class HandleT, HandleKind> friend class AnyObjectHandleImpl;
+  struct AnyObjectHandleTag {};
+  ReferenceBase(AnyObjectHandleTag, ReferenceBase Other)
+      : ReferenceBase(Other) {}
+
+  uint64_t InternalRef;
+
+#if LLVM_ENABLE_ABI_BREAKING_CHECKS
+  const CASDB *CAS;
+  bool IsHandle;
+#endif
```

Comment:
The main reason I added the `bool IsHandle` was to give that info when printing (the assertion this supported was just extra, and unnecessary because the `operator==()` overloads avoid this happening). I realized we really *always* want that info, and can provide it since there's always a concrete class to pass in. I dropped the `bool` entirely in the update / f766105ab52ea8cd9dd828ce2b65c49b219305b9.